<a href="https://colab.research.google.com/github/kaylatucker/CS690_Capstone/blob/main/03_notebooks/05_customer_experience.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Customer experience

In [ ]:
!git clone https://github.com/kaylatucker/CS690_Capstone.git

Cloning into 'CS690_Capstone'...
remote: Enumerating objects: 188, done.
remote: Counting objects: 100% (148/148), done.
remote: Compressing objects: 100% (137/137), done.
remote: Total 188 (delta 82), reused 23 (delta 11), pack-reused 40 (from 4)
Receiving objects: 100% (188/188), 84.71 MiB | 9.07 MiB/s, done.
Resolving deltas: 100% (84/84), done.
Updating files: 100% (39/39), done.


In [ ]:
from pathlib import Path
import pandas as pd

REPO_DIR = Path("/content/CS690_Capstone")
PROCESSED_DIR = REPO_DIR / "02_data" / "processed"
RAW_DIR = REPO_DIR / "02_data" / "raw"

print("Repo exists:", REPO_DIR.exists())

Repo exists: True


### Load what's needed

In [ ]:
orders = pd.read_csv(
    PROCESSED_DIR / "orders_clean.csv",
    parse_dates=[
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
)

order_summary = pd.read_csv(
    PROCESSED_DIR / "order_item_summary.csv"
)

reviews = pd.read_csv(
    PROCESSED_DIR / "order_reviews_clean.csv",
    parse_dates=[
        "review_creation_date",
        "review_answer_timestamp"
    ]
)

products = pd.read_csv(
    PROCESSED_DIR / "products_clean.csv"
)

seller_metrics = pd.read_csv(
    PROCESSED_DIR / "seller_metrics.csv",
    dtype={"seller_zip_code_prefix": str}
)

items = pd.read_csv(
    PROCESSED_DIR / "order_items_clean.csv",
    parse_dates=["shipping_limit_date"]
)

customers = pd.read_csv(
    RAW_DIR / "olist_customers_dataset.csv",
    dtype={"customer_zip_code_prefix": str}
)

In [ ]:
print("Orders:", orders.shape)
print("Order summary:", order_summary.shape)
print("Reviews:", reviews.shape)
print("Products:", products.shape)
print("Seller metrics:", seller_metrics.shape)
print("Items:", items.shape)
print("Customers:", customers.shape)

Orders: (99441, 11)
Order summary: (98666, 5)
Reviews: (99224, 7)
Products: (32951, 10)
Seller metrics: (3095, 16)
Items: (112650, 7)
Customers: (99441, 5)


### reducing reviews to one per order again



In [ ]:
order_reviews = (
    reviews
    .groupby("order_id", as_index=False)
    .agg(
        review_score=("review_score", "mean"),
        review_count=("review_score", "count")
    )
)

print("Order review rows:", len(order_reviews))
print("Unique order IDs:", order_reviews["order_id"].nunique())

Order review rows: 98673
Unique order IDs: 98673


## Creating order-level product/category characteristics without multiplying rows

### Join item rows to product categories

In [ ]:
items_products = items.merge(
    products[
        [
            "product_id",
            "product_category_name",
            "product_category_name_english"
        ]
    ],
    on="product_id",
    how="left",
    validate="many_to_one"
)

print("Rows after item-product join:", len(items_products))

Rows after item-product join: 112650


### Summarize product/category characteristics to one row per order

In [ ]:
order_product_summary = (
    items_products
    .groupby("order_id", as_index=False)
    .agg(
        product_count=("product_id", "nunique"),
        category_count=("product_category_name_english", "nunique")
    )
)

print("Order product summary rows:", len(order_product_summary))
print("Unique order IDs:", order_product_summary["order_id"].nunique())

display(order_product_summary.head())

Order product summary rows: 98666
Unique order IDs: 98666


,order_id,product_count,category_count
0,00010242fe8c5a6d1ba2dd792cb16214,1,1
1,00018f77f2f0320c557190d7a144bdd3,1,1
2,000229ec398224ef6ca0657da4fc703e,1,1
3,00024acbcdf0a6daa1e931b038114c75,1,1
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,1


### creating a readable primary category for orders where possible

In [ ]:
order_category = (
    items_products
    .groupby("order_id")["product_category_name_english"]
    .agg(
        lambda x: x.dropna().mode().iloc[0]
        if not x.dropna().mode().empty
        else pd.NA
    )
    .reset_index(name="primary_category")
)

print("Order category rows:", len(order_category))

Order category rows: 98666


In [ ]:
order_category.head()

,order_id,primary_category
0,00010242fe8c5a6d1ba2dd792cb16214,cool_stuff
1,00018f77f2f0320c557190d7a144bdd3,pet_shop
2,000229ec398224ef6ca0657da4fc703e,furniture_decor
3,00024acbcdf0a6daa1e931b038114c75,perfumery
4,00042b26cf59d7ce69dfabb4e55b4fd9,garden_tools


## Seller characteristics at the order level

In [ ]:
# Identify number of sellers per order
order_sellers = (
    items
    .groupby("order_id")
    .agg(
        seller_count=("seller_id", "nunique"),
        seller_id=("seller_id", "first")
    )
    .reset_index()
)

# Seller ID is only unambiguous for single-seller orders
order_sellers.loc[
    order_sellers["seller_count"] > 1,
    "seller_id"
] = pd.NA

print("Order-seller rows:", len(order_sellers))
print("Unique order IDs:", order_sellers["order_id"].nunique())

print("\nSeller count distribution:")
display(
    order_sellers["seller_count"]
    .value_counts()
    .sort_index()
    .to_frame("orders")
)

Order-seller rows: 98666
Unique order IDs: 98666

Seller count distribution:


,orders
seller_count,
1,97388
2,1219
3,54
4,3
5,2


### Attaching selected seller characteristics

In [ ]:
seller_features = seller_metrics[
    [
        "seller_id",
        "seller_order_count",
        "seller_items_sold",
        "seller_total_sales",
        "seller_avg_item_value",
        "seller_product_count",
        "seller_category_count",
        "seller_avg_freight",
        "seller_zip_code_prefix",
        "seller_city",
        "seller_state"
    ]
].copy()

order_seller_summary = order_sellers.merge(
    seller_features,
    on="seller_id",
    how="left",
    validate="many_to_one"
)

print("Order-seller summary rows:", len(order_seller_summary))
print(
    "Orders with seller characteristics:",
    order_seller_summary["seller_order_count"].notna().sum()
)

Order-seller summary rows: 98666
Orders with seller characteristics: 97388


### Customer geography

In [ ]:
customer_features = customers[
    [
        "customer_id",
        "customer_unique_id",
        "customer_zip_code_prefix",
        "customer_city",
        "customer_state"
    ]
].copy()

print("Customer rows:", len(customer_features))
print("Unique customer IDs:", customer_features["customer_id"].nunique())

Customer rows: 99441
Unique customer IDs: 99441


In [ ]:
customer_experience = (
    orders
    .merge(
        order_reviews,
        on="order_id",
        how="left",
        validate="one_to_one"
    )
    .merge(
        order_summary,
        on="order_id",
        how="left",
        validate="one_to_one"
    )
    .merge(
        order_product_summary,
        on="order_id",
        how="left",
        validate="one_to_one"
    )
    .merge(
        order_category,
        on="order_id",
        how="left",
        validate="one_to_one"
    )
    .merge(
        order_seller_summary,
        on="order_id",
        how="left",
        validate="one_to_one"
    )
    .merge(
        customer_features,
        on="customer_id",
        how="left",
        validate="many_to_one"
    )
)

#### validate the grain

In [ ]:
print("Customer-experience rows:", len(customer_experience))
print(
    "Unique order IDs:",
    customer_experience["order_id"].nunique()
)
print(
    "Duplicate order IDs:",
    customer_experience["order_id"].duplicated().sum()
)

display(customer_experience.head())

Customer-experience rows: 99441
Unique order IDs: 99441
Duplicate order IDs: 0


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,delivery_days,delivery_delay_days,...,seller_product_count,seller_category_count,seller_avg_freight,seller_zip_code_prefix,seller_city,seller_state,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,8.436574,-7.107488,...,14.0,6.0,12.969245,09350,maua,SP,7c396fd4830fd04220f754e42b4e5bff,03149,sao paulo,SP
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,13.782037,-5.355729,...,25.0,4.0,16.621746,31570,belo horizonte,SP,af07308b275d755c9edb36a90c618231,47813,barreiras,BA
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,9.394213,-17.245498,...,95.0,10.0,17.446427,14840,guariba,SP,3a653a41f6f9fc3d2a113cf8398680e8,75265,vianopolis,GO
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15,13.208750,-12.980069,...,20.0,1.0,20.633141,31842,belo horizonte,MG,7c142cf63193a1473d2e66489a9ae977,59296,sao goncalo do amarante,RN
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26,2.873877,-9.238171,...,15.0,5.0,14.331092,08752,mogi das cruzes,SP,72632f0f9dd73dfee390c9b22eb56dd6,09195,santo andre,SP


## Validation check

In [ ]:
print("Rows:", len(customer_experience))
print("Unique orders:", customer_experience["order_id"].nunique())
print("Duplicate orders:", customer_experience["order_id"].duplicated().sum())

print("\nMissing review scores:",
      customer_experience["review_score"].isna().sum())

print("Orders with no item summary:",
      customer_experience["order_value"].isna().sum())

print("Orders with no seller count:",
      customer_experience["seller_count"].isna().sum())

print("Multi-seller orders:",
      (customer_experience["seller_count"] > 1).sum())

print("Orders with seller characteristics:",
      customer_experience["seller_order_count"].notna().sum())

Rows: 99441
Unique orders: 99441
Duplicate orders: 0

Missing review scores: 768
Orders with no item summary: 775
Orders with no seller count: 775
Multi-seller orders: 1278
Orders with seller characteristics: 97388


In [ ]:
analysis_fields = [
    "review_score",
    "delivery_days",
    "delivery_delay_days",
    "late_delivery_flag",
    "order_value",
    "freight_total",
    "freight_percentage",
    "items_per_order",
    "product_count",
    "category_count",
    "primary_category",
    "seller_count",
    "seller_order_count",
    "seller_total_sales",
    "seller_avg_item_value",
    "seller_product_count",
    "seller_category_count",
    "seller_avg_freight",
    "customer_state",
    "seller_state"
]

display(
    customer_experience[analysis_fields]
    .isna()
    .sum()
    .sort_values(ascending=False)
    .to_frame("missing_count")
)

,missing_count
delivery_days,2965
delivery_delay_days,2965
late_delivery_flag,2965
primary_category,2164
seller_product_count,2053
seller_state,2053
seller_avg_freight,2053
seller_category_count,2053
seller_total_sales,2053
seller_avg_item_value,2053


# Notes

- Built an order-level customer-experience dataset with one row per order.
- Final table contains 99,441 unique orders with no duplicate order IDs.
- `review_score` is the primary customer-experience outcome.
- Reviews were aggregated to one score per order before joining.
- Delivery measures include `delivery_days`, `delivery_delay_days`, and `late_delivery_flag`.
- Order-level transaction measures include order value, freight total, freight percentage, and items per order.
- Product characteristics include product count, category count, and a representative primary category.
- Seller characteristics are only assigned to single-seller orders to avoid ambiguous attribution.
- 1,278 multi-seller orders therefore have no seller-level characteristics.
- 775 orders have no item records and are missing item, product, and seller-derived measures.
- 768 orders have no review score.
- 2,965 orders are missing valid delivery-based measures because no customer delivery date is available.
- Customer geography matched successfully for all orders.
- Missing values were retained where they reflect unavailable or structurally ambiguous information rather than being artificially filled.

# Export

In [ ]:
OUTPUT_FILE = PROCESSED_DIR / "customer_experience.csv"

customer_experience.to_csv(
    OUTPUT_FILE,
    index=False
)

print("Saved to:", OUTPUT_FILE)

Saved to: /content/CS690_Capstone/02_data/processed/customer_experience.csv


## Verify

In [ ]:
customer_experience_check = pd.read_csv(
    OUTPUT_FILE,
    dtype={
        "customer_zip_code_prefix": str,
        "seller_zip_code_prefix": str
    }
)

print("Rows:", len(customer_experience_check))
print(
    "Unique orders:",
    customer_experience_check["order_id"].nunique()
)
print(
    "Duplicate orders:",
    customer_experience_check["order_id"].duplicated().sum()
)

Rows: 99441
Unique orders: 99441
Duplicate orders: 0


In [ ]:
!git config --global user.email "ktuck011@odu.edu"
!git config --global user.name "Kayla Tucker"

In [ ]:
%cd /content/CS690_Capstone

!git status --short
!git add 02_data/processed/customer_experience.csv
!git commit -m "Add customer experience dataset"

/content/CS690_Capstone
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean


In [ ]:
!git push

fatal: could not read Username for 'https://github.com': No such device or address


In [ ]:
from google.colab import userdata
import subprocess

repo = "/content/CS690_Capstone"
token = userdata.get("GITHUB_TOKEN")

branch = subprocess.check_output(
    ["git", "-C", repo, "branch", "--show-current"],
    text=True
).strip()

result = subprocess.run(
    [
        "git",
        "-C", repo,
        "push",
        f"https://x-access-token:{token}@github.com/kaylatucker/CS690_Capstone.git",
        f"HEAD:{branch}"
    ],
    text=True
)

print("Push return code:", result.returncode)

Push return code: 0


AI Assistance Statement
Google Gemini was used to assist with drafting and formatting some Markdown cells in this notebook. I reviewed and edited the generated text before including it. I also used Google Colab’s code-completion/autocomplete feature when it suggested portions of code that matched what I intended to write. All code, analyses, and results were reviewed and verified by me.

Google. (2026). Gemini [Large language model]. https://gemini.google.com/
